### PR TITLE
Add initial schema and automatic interface generation

### PR DESCRIPTION
**Pull request by bruno-f-cruz**
This PR:
 - Refactors the current repository directory structure in accordance with the latest agreed upon harptech specs;
 - Adds a first schema version for the Harp Output Expander device (device.yml) and the necessary .Net generators to build the high-level bonsai interface.

### CODE REVIEW

**Comment by filcarv** [comment]
**Hardware/Mechanical**'s **README** file should be updated to Copyright Filipe Carvalho 2020 **-2023**

The Firmware solution name should be renamed to **OutputExpander** in order to maintain coherence with the other repos.

The InputExpander project of Atmel Studio is still available in Solution Explorer. Should be removed.

**Comment by bruno-f-cruz** [comment]
> Hardware/Mechanical's README file should be updated to Copyright Filipe Carvalho 2020 -2023
Done

> The Firmware solution name should be renamed to OutputExpander in order to maintain coherence with the other repos.
> The InputExpander project of Atmel Studio is still available in Solution Explorer. Should be removed.
As I am not super comfortable editing these solutions files, can you change them and PR into my branch? thanks!

File: device.yml [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+      Pwm0ToOut1: 0x1
+      Pwm0ToOut2: 0x2
+      Pwm0ToOut3: 0x4
+      Pwm1ToOut6: 0x8
+      Pwm1ToOut7: 0x10
+      Pwm1ToOut8: 0x20
+      Pwm2ToOut9: 0x40
+      Stim0ToOut0: 0x80
+      Stim0ToOut5: 0x100
+  Pwm:
```
Comment (glopesdev) [inline device.yml:357]: Consider renaming to `PwmChannels`

File: device.yml [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+    description: Specifies the available PWM protocols.
+    bits:
+      Pwm0: 0x1
+      Pwm1: 0x2
+      Pwm2: 0x4
+  Stim:
```
Comment (glopesdev) [inline device.yml:363]: Consider renaming to `StimChannels`

File: device.yml [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+      Software: 0
+      Aux0Rising: 1
+      Aux0Falling: 2
+      Aux0WhileHigh: 3
+      Aux0WhileLow: 4
+      Aux1Rising: 16
+      Aux1Falling: 32
+      Aux1WhileHigh: 64
+      Aux1WhileLow: 128
+  EnableFlag:
```
Comment (glopesdev) [inline device.yml:385]: This can be removed as it is built into the Harp package

File: device.yml [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+    description: The state of a configuration.
+    values:
+      Disabled: 0
+      Enabled: 1
+  ExpansionBoardTypes:
```
Comment (glopesdev) [inline device.yml:390]: Rename to `ExpansionBoardType`

File: device.yml [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+      Encoder: 1
+      ServoMotor1: 2
+      ServoMotor2: 3
+      ServoMotor3: 4
+      OpticalFlow: 5
+  EncoderSamplingRate:
+    description: Specifies the sampling rate of the encoder.
+    values:
+      At50Hz: 0
+      At00Hz: 1
```
Comment (glopesdev) [inline device.yml:403]: Rename to `SamplingRate100Hz` (or consistent with above)

File: device.yml [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+      Breakout: 0
+      Encoder: 1
+      ServoMotor1: 2
+      ServoMotor2: 3
+      ServoMotor3: 4
+      OpticalFlow: 5
+  EncoderSamplingRate:
+    description: Specifies the sampling rate of the encoder.
+    values:
+      At50Hz: 0
```
Comment (glopesdev) [inline device.yml:402]: Consider alternatives to `At`:
  - [x] `SamplingRate50Hz`
  - [ ] `Sampling50Hz`
  - [ ] `Fs50Hz` / `FS50Hz`

File: device.yml [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+  PwmStop:
+    address: 67
+    type: U8
+    access: Write
+    maskType: Pwm
+    description: Stops the a PWM on the specified channels.
+  PwmRiseEventEnable:
+    address: 68
+    type: U8
+    access: Write
```
Comment (glopesdev) [inline device.yml:185]: Add missing `Event` access

File: device.yml [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+    access: [Write, Event]
+    maskType: Pwm
+    description: Starts the a PWM on the specified channels. An event will be generated if the start was triggered by an auxiliary input.
+  PwmStop:
+    address: 67
+    type: U8
+    access: Write
+    maskType: Pwm
+    description: Stops the a PWM on the specified channels.
+  PwmRiseEventEnable:
```
Comment (glopesdev) [inline device.yml:182]: Consider renaming to `PwmRiseEvent`

File: device.yml [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+    <<: *outpulsewidth
+    address: 85
+  Out9PulseWidth:
+    <<: *outpulsewidth
+    address: 86
+  ExpansionBoard:
+    address: 87
+    type: U8
+    access: Write
+    maskType: ExpansionBoardTypes
```
Comment (glopesdev) [inline device.yml:267]: Rename to `ExpansionBoardType`

File: device.yml [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+  Encoder:
+    address: 90
+    type: U16
+    access: Event
+    description: Generated event with the latest read from the magnetic encoder.
+  EncoderSampling:
+    address: 91
+    type: U8
+    access: Write
+    maskType: EncoderSamplingRate
```
Comment (glopesdev) [inline device.yml:287]: Consider renaming the mask type to `EncoderSamplingRateMode`

**Review by glopesdev — CHANGES_REQUESTED** [review]

**Review by glopesdev — APPROVED** [review]

**Pull request merged**